Issue 15750: PIP-105: Store Subscription properties

eolivelli opened the pull request:
Fixes #15750

### Motivation

The Subscription properties are not properly stored on MetadataService

### Modifications

Store the subscription properties on the MetadataService (ManagedCursorInfo)
### Verifying this change

This change is already covered by existing tests, I have only added some additional assertions

eolivelli commented:
Even if we are changing some .proto file I believe that we can cherry-pick this change to branch-2.10, in order to unblock PIP-105. This feature wasn't working, so we are not going to break anything.

codelipenghui reviewed (commented):
How do the consumers change the subscription properties?
The change looks like the properties will not change after the cursor is created.

The previous behavior is the subscription will apply the properties from the first connected consumer.

dlg99 commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:
```diff
@@ -361,6 +369,18 @@ public void operationComplete(ManagedCursorInfo info, Stat stat) {
+                Map<String, String> recoveredCursorProperties = Collections.emptyMap();
+                if (info.getCursorPropertiesCount() > 0) {
+                    // Recover properties map
+                    recoveredCursorProperties = Maps.newHashMap();
```
nit: in 4 lines here (and in some other places in the change) we have two ways of creating new map (`Collections.emptyMap()` and `Maps.newHashMap()`).
While nothing is wrong with either one having both requires reviewer to think about motivation for having two, differences, etc. Do we really need both?

dlg99 replied:
Exactly my point ;) Can't guess that without looking at the library code or docs.
"ImmutableMap.of()" can express this in a more visible way.

dlg99 commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:
```diff
@@ -845,11 +845,12 @@ public ManagedCursor openCursor(String cursorName) throws InterruptedException,
     public ManagedCursor openCursor(String cursorName, InitialPosition initialPosition)
             throws InterruptedException, ManagedLedgerException {
-        return openCursor(cursorName, initialPosition, Collections.emptyMap());
+        return openCursor(cursorName, initialPosition, Collections.emptyMap(), Collections.emptyMap());
```
we may be better of with something like `final static Map<blah, blah> = Maps.ImmutableMap();` as a field to avoid creating empty maps adding to short-lived memory garbage.

eolivelli replied:
Same as above

dlg99 reviewed (approved):
LGTM, couple of nitpicks.

eolivelli commented:
> How do the consumers change the subscription properties? The change looks like the properties will not change after the cursor is created.
> 
> The previous behavior is the subscription will apply the properties from the first connected consumer.

You are right.
I have a follow up patch in which I add support for updating the cursorProperties. 
I will port part of those changes here in order to persist the properties populated from the first Consumer that sets a non empty property set.

Honestly I don't like that much that behavior and I would prefer  to let the user modify the properties using an explicit PulsarAdmin function (see #15751 )

eolivelli commented:
@codelipenghui I have updated this patch.
Now the behaviour is 100% backward compatible

eolivelli reviewed (commented):
@codelipenghui I have updated the patch.
@dlg99 At the moment I have used the same utilities used in the files, I believe that we should decide the overall style for the things you suggested and apply the changes globally

nicoloboschi commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:
```diff
@@ -313,6 +316,52 @@ public Map<String, Long> getProperties() {
+
+                    @Override
+                    public void operationFailed(MetaStoreException e) {
+                        log.info("[{}] Error while updating ledger cursor: {} properties {}", ledger.getName(),
```
log.error ?

eolivelli replied:
good catch

nicoloboschi commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:
```diff
@@ -313,6 +316,52 @@ public Map<String, Long> getProperties() {
+
+            @Override
+            public void operationFailed(MetaStoreException e) {
+                log.info("[{}] Error while updating ledger cursor: {} properties {}", ledger.getName(),
```
log.error ?

nicoloboschi commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:
```diff
@@ -2619,6 +2683,20 @@ private List<LongProperty> buildPropertiesMap(Map<String, Long> properties) {
+            return Collections.emptyList();
+        }
+
+        List<StringProperty> stringProperties = Lists.newArrayList();
```
nit: we can create a list with the same size of properties#size

nicoloboschi commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/nonpersistent/NonPersistentSubscription.java:
```diff
@@ -526,4 +526,15 @@ public void updateLastActive() {
+
+    @Override
+    public CompletableFuture<Void> updateSubscriptionProperties(Map<String, String> subscriptionProperties) {
+        if (subscriptionProperties == null) {
```
or empty ?

eolivelli replied:
there is not much difference

nicoloboschi commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java:
```diff
@@ -1094,6 +1094,20 @@ public Map<String, String> getSubscriptionProperties() {
+    @Override
+    public CompletableFuture<Void> updateSubscriptionProperties(Map<String, String> subscriptionProperties) {
+        Map<String, String> newSubscriptionProperties;
+        if (subscriptionProperties == null) {
```
or empty ?

eolivelli replied:
there is not much difference

nicoloboschi reviewed (changes requested):
Good work!

codelipenghui commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedCursor.java:
```diff
@@ -79,6 +81,20 @@ enum IndividualDeletedEntries {
+      * @param cursorProperties
+      * @return a handle to the result of the operation
+      */
+     default CompletableFuture<Void> updateCursorProperties(Map<String, String> cursorProperties) {
```
```suggestion
     default CompletableFuture<Void> setCursorProperties(Map<String, String> cursorProperties) {
```

codelipenghui commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java:
```diff
@@ -243,7 +243,8 @@ ManagedCursor openCursor(String name, InitialPosition initialPosition) throws In
      * @throws ManagedLedgerException
      */
-    ManagedCursor openCursor(String name, InitialPosition initialPosition, Map<String, Long> properties)
+    ManagedCursor openCursor(String name, InitialPosition initialPosition, Map<String, Long> properties,
+                             Map<String, String> ledgerProperties)
```
```suggestion
    ManagedCursor openCursor(String name, InitialPosition initialPosition, Map<String, Long> properties,
                             Map<String, String> cursorProperties)
```

codelipenghui commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:
```diff
@@ -898,21 +900,34 @@ public void openCursorComplete(ManagedCursor cursor, Object ctx) {
                 if (MapUtils.isEmpty(subscription.getSubscriptionProperties())
                         && MapUtils.isNotEmpty(subscriptionProperties)) {
-                    subscription.getSubscriptionProperties().putAll(subscriptionProperties);
+                    updateProperties = subscription
+                            .updateSubscriptionProperties(subscriptionProperties)
+                            .thenApply(___ -> _subscription);
```
I have left a comment here https://github.com/apache/pulsar/pull/15757#issuecomment-1136921453

eolivelli commented:
@nicoloboschi @codelipenghui  I have addressed your comments, PTAL 

@codelipenghui I think it is better to not change the behaviour for this patch

nicoloboschi reviewed (approved):
LGTM

eolivelli commented:
@codelipenghui , as discussed offline, I have changed the behavior of subscription properties, now they are like "initial position". You can set them only at creation time.
In my follow up patch #15751 I am adding support for updating them using PulsarAdmin

eolivelli commented:
/pulsar-bot rerun-failure-checks

eolivelli commented:
/pulsarbot rerun-failure-checks

lhotari commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java:
```diff
@@ -1094,6 +1093,20 @@ public Map<String, String> getSubscriptionProperties() {
+        }
+        return cursor.setCursorProperties(newSubscriptionProperties)
+                .thenRun(() -> {
+                    this.subscriptionProperties = newSubscriptionProperties;
```
Are there any thread safety considerations here? Would it be required to make subscriptionProperties a volatile field?

eolivelli replied:
good catch. I added volatile only on NonPersistentSubscription and I missed it here.
Fixed, thank you @lhotari

lhotari reviewed (commented):
I added a question about thread safety.

eolivelli commented:
/pulsarbot rerun-failure-checks

eolivelli commented:
/pulsarbot rerun-failure-checks

eolivelli commented:
/pulsarbot rerun-failure-checks

lhotari reviewed (approved):
LGTM. Good work @eolivelli

eolivelli commented:
/pulsar-bot rerun-failure-checks
